core/common/src/main/java/alluxio/resource/DynamicResourcePool.java

```diff
@@ -238,32 +241,37 @@ public DynamicResourcePool(Options options) {
 
     mMaxCapacity = options.getMaxCapacity();
     mMinCapacity = options.getMinCapacity();
-    mAvailableResources = new ArrayDeque<>(Math.min(mMaxCapacity, 32));
+    mAvailableResources = new ConcurrentLinkedDeque<>();
+    mCurrentCapacity = new AtomicInteger();
+    mWaitingThreadsNumber = new AtomicInteger();
 
     mGcFuture = mExecutor.scheduleAtFixedRate(() -> {
       List<T> resourcesToGc = new ArrayList<>();
 
-      try {
-        mLock.lock();
-        if (mResources.size() <= mMinCapacity) {
-          return;
-        }
-        int currentSize = mResources.size();
-        Iterator<ResourceInternal<T>> iterator = mAvailableResources.iterator();
-        while (iterator.hasNext()) {
-          ResourceInternal<T> next = iterator.next();
-          if (shouldGc(next)) {
-            resourcesToGc.add(next.mResource);
-            iterator.remove();
-            mResources.remove(next.mResource);
-            currentSize--;
-            if (currentSize <= mMinCapacity) {
-              break;
+      if (mResources.size() <= mMinCapacity) {
+        return;
+      }
+      ResourceInternal<T> queueTailResource;
+      while ((queueTailResource = mAvailableResources.pollLast()) != null) {
+        if (shouldGc(queueTailResource)) {
+          resourcesToGc.add(queueTailResource.mResource);
+          remove(queueTailResource.mResource);
+          if (mCurrentCapacity.get() <= mMinCapacity) {
+            break;
+          }
+        } else {
+          // add the resource back the deque
+          mAvailableResources.offerLast(queueTailResource);
+          if (mWaitingThreadsNumber.get() > 0) {
```

Comment:
Since the lock is no longer acquired, there is no longer consistent access between multiple variables. I see various places where decisions are made from one atomic int, that affects/mutates something else. Will that become a problem, since now those accesses are no longer consistent (within a lock)?

Reply:
> Since the lock is no longer acquired, there is no longer consistent access between multiple variables. I see various places where decisions are made from one atomic int, that affects/mutates something else. Will that become a problem, since now those accesses are no longer consistent (within a lock)?

Thanks @gpang, the purpose of this change is to make sure these variables are consistent without mLock when pool is not full or there is not waiting thread, so in most times it's lock-free. 

Combination of mCurrentCapacity and mAvailableResources is just like mCurrentCapacity and mResources in alluxio.resource.ResourcePool, so it's should be easier to understand.

I suspect your question is mainly about mWaitingThreadsNumber, which impacts whether to signal waiting threads. I think there are two corner cases when thread A calling accquire method runs to line 335 - 345.

1. A is not permitted to allocated new resource and has not yet increase mWaitingThreadsNumber.
2. A has increased mWaitingThreadsNumber, entered while loop, but has not called await.

In case 1, if another thread B releases resource, A will find the resource when running into the loop (line 341)
In case 2, if anothere thead B releases resource, A will not be notified because it's not waiting, so it will continue to call await and find the new resource in next loop. 

So the worst case is thread A throws TimeoutException even if B releases resource, and this case should be rare to happen.

-----------------------------

Below are details about the important variables and scenarios.

**variables**

- _mCurrentCapacity_ - Is AtomicInteger, used as a resource counter and permit token, only thread successfully increment this can allocate a new resource.
- _mAvailableResources_ - Is thread-safe it self, used to contain resource that are released and not yet gc.
- _mLock_ - when pool is full, this lock is used by user thread to await, or by user thread / gc thread to notify available resource.
- _mWaitingThreadsNumber_ - Is AtomicInteger, used as a counter of waiting threads, if it's 0, we don't need to notify when resource is available.

**status changing scenarios**

- _accquire_ - poll from mAvailableResources; if don't get a resource, trying to allocate new resource by trying to increase mCurrentCapacity; if not permitted, increase / release mWaitingThreadsNumber in a try / finally block and trying to await until be notified or timeout.
- _release_ - add resource to mAvailableResources, then if there is wating threads, call mLock to wake one thread
- _gc_ - remove timeout resources from mAvailableResources's queue end (oldest) and reduce mCurrentCapacity in a loop, until the queue is empty, or we removed a not-timeout resource, which should be added back to queue just like _release_ a resource and signal waiting thread if needed.